src/AppInstallerRepositoryCore/Public/winget/PinningData.h

```diff
@@ -0,0 +1,105 @@
+#include <vector>
+
+namespace AppInstaller::Repository::Microsoft
+{
+    struct PinningIndex;
+}
+
+namespace AppInstaller::Pinning
+{
+    // Possible ways to consider pins when getting a package's available versions
+    enum class PinBehavior
+    {
+        // Ignore pins, returns all available versions.
+        IgnorePins,
+        // Include available versions for packages with a Pinning pin.
+        // Blocking pins and Gating pins still respected.
+        IncludePinned,
+        // Respect all the types of pins.
+        ConsiderPins,
+    };
+
+    // The public representation of the pinning database.
+    struct PinningData
+    {
+        // Creates an empty pinning data.
+        PinningData();
+
+        // Enum to make the pinning data disposition clear in the caller.
+        enum class Disposition
+        {
+            // The data can only be read.
+            ReadOnly,
+            // The data can be read and written.
+            ReadWrite,
+        };
+
+        // Creates a usable pinning data with the given read/write capability.
+        PinningData(Disposition disposition);
+
+        PinningData(const PinningData&);
+        PinningData& operator=(const PinningData&);
+        PinningData(PinningData&&) noexcept;
+        PinningData& operator=(PinningData&&) noexcept;
+        ~PinningData();
+
+        // Determines if the pinning database is opened
+        operator bool() const;
+        bool IsDatabaseConnected() const;
+
+        // Pass through functions to the index itself
+        void AddOrUpdatePin(const Pin& pin);
+        void RemovePin(const PinKey& pinKey);
+        std::optional<Pin> GetPin(const PinKey& pinKey);
+        std::vector<Pin> GetAllPins();
+        bool ResetAllPins(std::string_view sourceId = {});
+
+        // A type used for evaluating the pinning state for a given package.
+        struct PinStateEvaluator
+        {
+            PinStateEvaluator(
+                PinBehavior behavior,
+                std::shared_ptr<AppInstaller::Repository::Microsoft::PinningIndex> database,
+                const std::shared_ptr<AppInstaller::Repository::IPackageVersion>& installedVersion);
+
+            PinStateEvaluator(const PinStateEvaluator&);
+            PinStateEvaluator& operator=(const PinStateEvaluator&);
+            PinStateEvaluator(PinStateEvaluator&&) noexcept;
+            PinStateEvaluator& operator=(PinStateEvaluator&&) noexcept;
+
+            ~PinStateEvaluator();
+
+            // Gets the latest available package that fits within the pinning restrictions.
+            std::shared_ptr<AppInstaller::Repository::IPackageVersion> GetLatestAvailableVersionForPins(const std::shared_ptr<AppInstaller::Repository::IPackage>& package);
+
+            // Determines if the given version is an update to the installed version that this object was created with.
+            bool IsUpdate(const std::shared_ptr<AppInstaller::Repository::IPackageVersion>& availableVersion);

```

Comment:
These functions assume that the available version passed as parameters are for the same package as the installed version, right? If so, I'd mention that in a comment